Feature/ms-3787-readfilter (#3787, filtering of search results by read status)

extracts opened the pull request:
PR for [#3787](https://projects.effective-webwork.de/projects/commsy/work_packages/3787) which enables filtering by read status (all, new only, modified only, read only).

The read status for the user's searchable items gets cached so that it doesn't need to be computed again every time.

Also, the read status (new/changed) now gets displayed as an exclamation mark icon next to each search result.

Note that (for testing/debugging purposes) my branch for #3787 (`feature/ms-3787-readfilter`) is based on the `feature/3748_2` branch, so this PR is against that branch (so that it only displays my new commits). I don't intend to perform the actual merge with this PR. It's just to gather review feedback.

extracts committed:
just some code rewrite w/o any changes in function or logic

extracts committed:
#3787: initial WIP changes to support filtering of search results by the (user-specific) read status (new/changed/seen)

extracts committed:
#3787: implements filtering of search results by the (user-specific) read status (new/changed)

TODO: the "seen" (aka "read") option of the read status filter dropdown hasn't been implemented yet

extracts committed:
#3787: when filtering search results by read status, choosing "New" will now also find items with status "new_annotation", and choosing "changed" will also find items with status "changed_annotation"

extracts committed:
#3787: implements filtering search results by read status "Read" (aka "seen")

extracts committed:
moves generic logic from ReadStatusFilterCondition to a ReaderService method

extracts committed:
#3787: when filtering search results by read status, we now cache the read status of all processed items which speeds up subsequent retrieval

TODO: properly invalidate a cache item when its real item gets updated or marked as read

extracts committed:
#3787: fixes an issue where topics, groups (and institutions) weren't displayed when filtering search results by read status (new/changed/seen)

extracts committed:
#3787: avoids an exception related to items of type "institution" when filtering search results by read status (new/changed/seen)

extracts committed:
#3787: now updates the read status cache when an item gets marked as read

this fixes an issue when filtering search results by read status (new/changed/seen) where an outdated read status cache could cause search results to get displayed within the wrong filter group

extracts committed:
#3787: extracts the logic which returns a cached read status for a given item & user into its own method

extracts committed:
#3787: similar to any other list views in CommSy, search results now display a red or orange exclamation mark next to the modifier icon to indicate their read status (new/changed/new annotation/changed annotation)

EofChris commented on config/services.yaml:
```diff
@@ -1306,6 +1306,9 @@ services:
     App\Event\UserStatusChangedEvent:
         autowire: false
 
+    App\Event\ReadStatusWillChangeEvent:
```
Maybe use Pre and Post for naming?

```suggestion
    App\Event\ReadStatusPreChangeEvent:
```

EofChris commented on src/Event/ReadStatusWillChangeEvent.php:
```diff
@@ -0,0 +1,67 @@
+ * Class ReadStatusWillChangeEvent
+ * @package App\Event
+ */
+class ReadStatusWillChangeEvent extends Event
```
```suggestion
class ReadStatusPreChangeEvent extends Event
```

EofChris commented on legacy/classes/cs_reader_manager.php:
```diff
@@ -242,6 +242,15 @@ public function markItemsAsRead(array $itemIds, int $versionId, array $userIds =
+               /** @var \Symfony\Component\EventDispatcher\EventDispatcher $eventDispatcher */
+               $eventDispatcher = $symfonyContainer->get('event_dispatcher');
+
+               $readStatusWillChangeEvent = new \App\Event\ReadStatusWillChangeEvent($userId, $itemId, 'seen');
```
use constants

```suggestion
               $readStatusWillChangeEvent = new \App\Event\ReadStatusWillChangeEvent($userId, $itemId, App\Utils\ReaderService::READ_STATUS_SEEN);
```

EofChris commented on src/Event/ReadStatusWillChangeEvent.php:
```diff
@@ -0,0 +1,67 @@
+        $this->userId = $userId;
+        $this->itemId = $itemId;
+        $this->newReadStatus = $newReadStatus;
+    }
```
Consider some checks like:

```
       if (
           $limit !== self::SHOW_ENTRIES_ONLY_ACTIVATED &&
           $limit !== self::SHOW_ENTRIES_ONLY_DEACTIVATED &&
           $limit !== self::SHOW_ENTRIES_ACTIVATED_DEACTIVATED
       ) {
           throw new InvalidArgumentException('unknown limit given');
       }
```

EofChris commented on src/EventSubscriber/ReadStatusSubscriber.php:
```diff
@@ -0,0 +1,58 @@
+    public static function getSubscribedEvents()
+    {
+        return [
+            ReadStatusWillChangeEvent::class => 'onReadStatusWillChange',
```
```suggestion
            ReadStatusPreChangeEvent::class => 'ReadStatusPreChange',
```

EofChris commented on src/EventSubscriber/ReadStatusSubscriber.php:
```diff
@@ -0,0 +1,58 @@
+        ];
+    }
+
+    public function onReadStatusWillChange(ReadStatusWillChangeEvent $event)
```
```suggestion
    public function ReadStatusPreChange(ReadStatusPreChangeEvent $event)
```

EofChris commented on src/Utils/ItemService.php:
```diff
@@ -137,6 +145,42 @@ public function getItemFileList($itemId) {
+            CS_MATERIAL_TYPE,
+            CS_TODO_TYPE,
+            CS_USER_TYPE,
+            cs_userroom_item::ROOM_TYPE_USER,
```
```suggestion
            \cs_userroom_item::ROOM_TYPE_USER,
```

EofChris commented on src/Utils/ItemService.php:
```diff
@@ -2,14 +2,19 @@
-use Symfony\Component\Form\Form;
-
 use App\Services\LegacyEnvironment;
+use cs_userroom_item;
```
```suggestion
```

EofChris commented on src/Utils/ItemService.php:
```diff
@@ -137,6 +145,42 @@ public function getItemFileList($itemId) {
+     * @param integer[] $contextIds array of room IDs for rooms whose items shall be returned
+     * @return \cs_item[]
+     */
+    public function getSearchableItemsForContextIds(array $contextIds)
```
Oder Alternativ \cs_list

```suggestion
    public function getSearchableItemsForContextIds(array $contextIds): array
```

EofChris commented on src/Controller/SearchController.php:
```diff
@@ -729,13 +754,18 @@ private function prepareResults(TransformedPaginatorAdapter $searchResults, $cur
                     $status = $searchResult->getStatus();
                 }
+                if (method_exists($searchResult, 'getItemId')) {
+                    $item = $itemService->getItem($searchResult->getItemId());
```
Try using $searchResult here without the itemService

extracts replied:
I've not been successfull to use `$searchResult` instead of `$item`. I f I do so, I get:
```
Argument 1 passed to App\Utils\ReaderService::cachedReadStatusForItem() must be an instance of cs_item, instance of App\Entity\Discussions given, called in /var/www/html/src/Controller/SearchController.php on line 762
```

EofChris commented on src/Model/SearchData.php:
```diff
@@ -166,6 +171,24 @@ public function setAppearsInDescription(bool $appearsInDescription): SearchData
+     * @param string|null $selectedReadStatus
+     * @return SearchData
+     */
+    public function setSelectedReadStatus(?string $selectedReadStatus): SearchData
```
```suggestion
    public function setSelectedReadStatus(?string $selectedReadStatus): self
```

extracts committed:
#3787 review: renames the "ReadStatusWillChangeEvent" to "ReadStatusPreChangeEvent"

extracts committed:
#3787 review: more/better use of constants

extracts committed:
#3787 review: return "self" instead of the class name to better facilitate subclassing

extracts committed:
#3787: fixes an issue where filtering search results by read status could inadvertently return every searchable item in the database

this could happen if no items matched the chosen read status for the current user

extracts committed:
#3787: follow-up to 7e8fd6e which now (really) fixes an issue where filtering search results by read status could inadvertently return every searchable item in the database

this could happen if no items matched the chosen read status for the current user
